In [7]:
import pandas as pd
import numpy as np ,numpy.random
import matplotlib.pyplot as plt
import seaborn as sns
from dateutil.relativedelta import relativedelta
import random
from catboost import Pool, CatBoostRegressor, cv,CatBoostClassifier

# Table Rendements

In [8]:
url = "C:/Users/louis/Documents/Projet Finance Quantitative/"

In [1]:
database = pd.read_excel(url + "Universe_ESG.xlsx")
database = database.T.iloc[1:,]
database = database.reset_index()
database.columns = database.iloc[0,]
database = database.iloc[1:,]
database = database.rename(columns={"Name" : "Date"})
database["Date"] = pd.to_datetime(database["Date"])
Convertir = list(database.columns[1:])
database[Convertir] = database[Convertir].astype(float)
print(database.shape)
database.head(3)
pd.set_option('display.max_rows', 20)
db_rendements = database.copy()
db_rendements=db_rendements.set_index("Date")
db_rendements = db_rendements.stack()
db_rendements = pd.DataFrame(db_rendements,columns = {"Return"})
db_rendements = db_rendements.reset_index()
db_rendements = db_rendements.rename(columns = {0:"Titres"})
lag_1mois = db_rendements.sort_values(["Titres","Date"]).groupby("Titres")["Return"].shift(1)
lag_2mois = db_rendements.sort_values(["Titres","Date"]).groupby("Titres")["Return"].shift(2)
lag_3mois = db_rendements.sort_values(["Titres","Date"]).groupby("Titres")["Return"].shift(3)
lag_4mois = db_rendements.sort_values(["Titres","Date"]).groupby("Titres")["Return"].shift(4)
db_rendements["lag1"]=lag_1mois
db_rendements["lag2"]=lag_2mois
db_rendements["lag3"]=lag_3mois
db_rendements["lag4"]=lag_4mois
db_rendements.sort_values(["Titres","Date"])
db_rendements.iloc[400:405,:]

NameError: name 'pd' is not defined

##### Valeurs manquantes sur les rendements ?

In [2]:
db_rendements.isna().sum()

NameError: name 'db_rendements' is not defined

On peut observer qu'il n'y a pas de valeurs manquantes sur les données de rendement fournies initialement  

In [11]:
db_rendements.to_csv("db_rendements.csv", header=True,index=False)

In [12]:
database.to_csv("database.csv", header=True,index=False)

# Table de jointure

In [13]:
pd.set_option("display.max_rows",500)
Equivalence = pd.read_excel(url + "Universe_ESG.xlsx",usecols=["Ticker","Name"])
Equivalence['Abréviation'] = Equivalence['Ticker'].astype(str).str[:-7]
#Equivalence["Domaine"] = 
Equivalence

Ticker                          Name Abréviation
0    HES US Equity                     HESS CORP      HES US
1    NEM US Equity         NEWMONT GOLDCORP CORP      NEM US
2   INTC US Equity                    INTEL CORP     INTC US
3    IFF US Equity     INTL FLAVORS & FRAGRANCES      IFF US
4    MOS US Equity                 MOSAIC CO/THE      MOS US
5    EXC US Equity                   EXELON CORP      EXC US
6    KMB US Equity           KIMBERLY-CLARK CORP      KMB US
7    JNJ US Equity             JOHNSON & JOHNSON      JNJ US
8    BAX US Equity      BAXTER INTERNATIONAL INC      BAX US
9     FO US Equity                 FORD MOTOR CO       FO US
10    CC US Equity               CHEMOURS CO/THE       CC US
11   COP US Equity                CONOCOPHILLIPS      COP US
12   WFC US Equity              WELLS FARGO & CO      WFC US
13    WY US Equity               WEYERHAEUSER CO       WY US
14   TGT US Equity                   TARGET CORP      TGT US
15   MMM US Equity                         3M CO      MMM US
16    GM US Equity             GENERAL MOTORS CO       GM US
17   CMI US Equity                   CUMMINS INC      CMI US
18   HAS US Equity                    HASBRO INC      HAS US
19   DUK US Equity              DUKE ENERGY CORP      DUK US
20   EMN US Equity           EASTMAN CHEMICAL CO      EMN US
21    BK US Equity  BANK OF NEW YORK MELLON CORP       BK US
22   UFS US Equity                   DOMTAR CORP      UFS US
23   ECL US Equity                    ECOLAB INC      ECL US
24   SLB US Equity              SCHLUMBERGER LTD      SLB US
25   UPS US Equity    UNITED PARCEL SERVICE-CL B      UPS US
26   AT& US Equity                      AT&T INC      AT& US
27   NSC US Equity         NORFOLK SOUTHERN CORP      NSC US
28   PPL US Equity                      PPL CORP      PPL US
29    MO US Equity              ALTRIA GROUP INC       MO US
30   JLL US Equity        JONES LANG LASALLE INC      JLL US
31    CI US Equity                 CITIGROUP INC       CI US
32   ABT US Equity           ABBOTT LABORATORIES      ABT US
33   AMD US Equity        ADVANCED MICRO DEVICES      AMD US
34   CVX US Equity                  CHEVRON CORP      CVX US
35   CMA US Equity                  COMERICA INC      CMA US
36   DTE US Equity            DTE ENERGY COMPANY      DTE US
37   HSY US Equity                HERSHEY CO/THE      HSY US
38   KIM US Equity             KIMCO REALTY CORP      KIM US
39   NBL US Equity              NOBLE ENERGY INC      NBL US
40   IBM US Equity   INTL BUSINESS MACHINES CORP      IBM US
41  WELL US Equity                 WELLTOWER INC     WELL US
42    IR US Equity            INGERSOLL-RAND PLC       IR US
43   BKR US Equity               BAKER HUGHES CO      BKR US
44   WEC US Equity          WEC ENERGY GROUP INC      WEC US
45   OXY US Equity     OCCIDENTAL PETROLEUM CORP      OXY US
46   WMB US Equity              WILLIAMS COS INC      WMB US
47   UNP US Equity            UNION PACIFIC CORP      UNP US
48    WM US Equity          WASTE MANAGEMENT INC       WM US
49   CCL US Equity                 CARNIVAL CORP      CCL US

# Table Indicateurs

In [14]:
IND1 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="EBITDA_TO_REVENUE2")
IND1.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND2 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="CURRENT_EV_TO_T12M_EBITDA2")
IND2.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND3 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="EBITDA_to_Interest_Expense2")
IND3.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND4 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="TOT_DEBT_TO_EBITDA2")
IND4.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND5 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="BEST_EST_PE_NXT_YR2")
IND5.loc[0:116,:].fillna(int(-999),inplace=True)
IND5.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND6 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="PE_RATIO2")
IND6.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND7 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="DIVIDEND_INDICATED_YIELD2")
IND7.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND8 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="EQY_DVD_YLD_IND2")
IND8.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
IND9 = pd.read_excel(url + "Excel_fq_clean.xlsx",sheet_name="EPS_GROWTH2")
IND9.rename(columns={"F US Equity" : "FO US Equity","T US Equity" : "AT& US Equity","C US Equity" : "CI US Equity" },inplace=True)
indicateurs = IND1.add_suffix('_EBITDA_TO_REVENUE').merge(IND2.add_suffix('_CURRENT_EV_TO_T12M_EBITDA'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_CURRENT_EV_TO_T12M_EBITDA")
del indicateurs["Fonds_CURRENT_EV_TO_T12M_EBITDA"]
indicateurs = indicateurs.merge(IND3.add_suffix('_EBITDA_to_Interest_Expense'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_EBITDA_to_Interest_Expense")
del indicateurs["Fonds_EBITDA_to_Interest_Expense"]
indicateurs = indicateurs.merge(IND4.add_suffix('_TOT_DEBT_TO_EBITDA'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_TOT_DEBT_TO_EBITDA")
del indicateurs["Fonds_TOT_DEBT_TO_EBITDA"]
indicateurs = indicateurs.merge(IND5.add_suffix('_BEST_EST_PE_NXT_YR2'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_BEST_EST_PE_NXT_YR2")
del indicateurs["Fonds_BEST_EST_PE_NXT_YR2"]
indicateurs = indicateurs.merge(IND6.add_suffix('_PE_RATIO'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_PE_RATIO")
del indicateurs["Fonds_PE_RATIO"]
indicateurs = indicateurs.merge(IND7.add_suffix('_DIVIDEND_INDICATED_YIELD'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_DIVIDEND_INDICATED_YIELD")
del indicateurs["Fonds_DIVIDEND_INDICATED_YIELD"]
indicateurs = indicateurs.merge(IND8.add_suffix('_EQY_DVD_YLD_IND'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_EQY_DVD_YLD_IND")
del indicateurs["Fonds_EQY_DVD_YLD_IND"]
indicateurs = indicateurs.merge(IND9.add_suffix('_EPS_GROWTH'),how = "left",left_on="Fonds_EBITDA_TO_REVENUE",right_on="Fonds_EPS_GROWTH")
del indicateurs["Fonds_EPS_GROWTH"]
indicateurs = indicateurs.rename(columns={"Fonds_EBITDA_TO_REVENUE" : "Dates"})
indicateurs.head(2)

C:\Users\louis\Anaconda3\lib\site-packages\pandas\core\frame.py:4244: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  **kwargs


Dates  HES US Equity_EBITDA_TO_REVENUE  \
0 2010-01-31                         5.871767   
1 2010-02-28                         5.871767   

   NEM US Equity_EBITDA_TO_REVENUE  INTC US Equity_EBITDA_TO_REVENUE  \
0                        16.904533                         19.417867   
1                        16.904533                         19.417867   

   IFF US Equity_EBITDA_TO_REVENUE  MOS US Equity_EBITDA_TO_REVENUE  \
0                           6.3521                        11.826033   
1                           6.3521                        11.826033   

   EXC US Equity_EBITDA_TO_REVENUE  KMB US Equity_EBITDA_TO_REVENUE  \
0                           15.841                           5.9152   
1                           15.841                           5.9152   

   JNJ US Equity_EBITDA_TO_REVENUE  BAX US Equity_EBITDA_TO_REVENUE  ...  \
0                        11.730967                         3.405067  ...   
1                        11.730967                         3.405067  ...   

   IBM US Equity_EPS_GROWTH  WELL US Equity_EPS_GROWTH  \
0                  5.653033                   -21.9363   
1                  5.653033                   -21.9363   

   IR US Equity_EPS_GROWTH  BKR US Equity_EPS_GROWTH  \
0                      NaN                       NaN   
1                      NaN                       NaN   

   WEC US Equity_EPS_GROWTH  OXY US Equity_EPS_GROWTH  \
0                 -2.754833                 66.666667   
1                 -2.754833                 66.666667   

   WMB US Equity_EPS_GROWTH  UNP US Equity_EPS_GROWTH  \
0                   12.6984                   13.8889   
1                   12.6984                   13.8889   

   WM US Equity_EPS_GROWTH  CCL US Equity_EPS_GROWTH  
0                   6.4516                   -1.9608  
1                   6.4516                   -1.9608  

[2 rows x 451 columns]

In [15]:
indicateurs.isna().sum()

Dates                                          0
HES US Equity_EBITDA_TO_REVENUE                2
NEM US Equity_EBITDA_TO_REVENUE                2
INTC US Equity_EBITDA_TO_REVENUE               5
IFF US Equity_EBITDA_TO_REVENUE                2
MOS US Equity_EBITDA_TO_REVENUE               11
EXC US Equity_EBITDA_TO_REVENUE                2
KMB US Equity_EBITDA_TO_REVENUE                2
JNJ US Equity_EBITDA_TO_REVENUE                2
BAX US Equity_EBITDA_TO_REVENUE                2
FO US Equity_EBITDA_TO_REVENUE                 2
CC US Equity_EBITDA_TO_REVENUE                50
COP US Equity_EBITDA_TO_REVENUE                2
WFC US Equity_EBITDA_TO_REVENUE              119
WY US Equity_EBITDA_TO_REVENUE                 2
TGT US Equity_EBITDA_TO_REVENUE                2
MMM US Equity_EBITDA_TO_REVENUE                2
GM US Equity_EBITDA_TO_REVENUE                 2
CMI US Equity_EBITDA_TO_REVENUE                2
HAS US Equity_EBITDA_TO_REVENUE                5
DUK US Equity_EBITDA

On observe que chaqu'un des indicateurs possède des variables manquantes. Pour aller même plus loin, chaque titre a des valeurs manquantes pour tous ces indicateurs. Cependant, il sera à noter qu'aucun indicateur n'est disponible pour les deux derniers mois de l'étude (Octobre et Novemebre 2019) par conséquent ces valeurs manquantes ne seront pas prises en compte. 
On a cependant beaucoup d'autres valeurs manquantes : 

In [16]:
Valeurs_manquantes = pd.DataFrame(indicateurs.loc[0:116,:].isna().sum(),columns=["NA"]).reset_index()
Valeurs_manquantes = Valeurs_manquantes[Valeurs_manquantes["NA"]!= 0].reset_index(drop=True)
Valeurs_manquantes["Nom"]=Valeurs_manquantes["index"].astype(str).str[0:3]
Valeurs_manquantes

index   NA  Nom
0            INTC US Equity_EBITDA_TO_REVENUE    3  INT
1             MOS US Equity_EBITDA_TO_REVENUE    9  MOS
2              CC US Equity_EBITDA_TO_REVENUE   48  CC 
3             WFC US Equity_EBITDA_TO_REVENUE  117  WFC
4             HAS US Equity_EBITDA_TO_REVENUE    3  HAS
5              BK US Equity_EBITDA_TO_REVENUE  117  BK 
6             AMD US Equity_EBITDA_TO_REVENUE    3  AMD
7             CMA US Equity_EBITDA_TO_REVENUE  117  CMA
8             BKR US Equity_EBITDA_TO_REVENUE   72  BKR
9      CC US Equity_CURRENT_EV_TO_T12M_EBITDA   63  CC 
10    WFC US Equity_CURRENT_EV_TO_T12M_EBITDA  117  WFC
11     GM US Equity_CURRENT_EV_TO_T12M_EBITDA    9  GM 
12    CMI US Equity_CURRENT_EV_TO_T12M_EBITDA    6  CMI
13     BK US Equity_CURRENT_EV_TO_T12M_EBITDA  117  BK 
14    CMA US Equity_CURRENT_EV_TO_T12M_EBITDA  117  CMA
15    BKR US Equity_CURRENT_EV_TO_T12M_EBITDA   90  BKR
16  INTC US Equity_EBITDA_to_Interest_Expense   12  INT
17   MOS US Equity_EBITDA_to_Interest_Expense    9  MOS
18    CC US Equity_EBITDA_to_Interest_Expense   63  CC 
19   WFC US Equity_EBITDA_to_Interest_Expense  117  WFC
20   HAS US Equity_EBITDA_to_Interest_Expense    3  HAS
21    BK US Equity_EBITDA_to_Interest_Expense  117  BK 
22    MO US Equity_EBITDA_to_Interest_Expense  117  MO 
23   JLL US Equity_EBITDA_to_Interest_Expense  117  JLL
24   AMD US Equity_EBITDA_to_Interest_Expense    3  AMD
25   CMA US Equity_EBITDA_to_Interest_Expense  117  CMA
26   BKR US Equity_EBITDA_to_Interest_Expense   72  BKR
27   OXY US Equity_EBITDA_to_Interest_Expense   21  OXY
28          INTC US Equity_TOT_DEBT_TO_EBITDA    3  INT
29           MOS US Equity_TOT_DEBT_TO_EBITDA    9  MOS
30            CC US Equity_TOT_DEBT_TO_EBITDA   57  CC 
31           WFC US Equity_TOT_DEBT_TO_EBITDA  117  WFC
32            GM US Equity_TOT_DEBT_TO_EBITDA    3  GM 
33           CMI US Equity_TOT_DEBT_TO_EBITDA    6  CMI
34           HAS US Equity_TOT_DEBT_TO_EBITDA    3  HAS
35            BK US Equity_TOT_DEBT_TO_EBITDA  117  BK 
36           AMD US Equity_TOT_DEBT_TO_EBITDA    3  AMD
37           CMA US Equity_TOT_DEBT_TO_EBITDA  117  CMA
38           BKR US Equity_TOT_DEBT_TO_EBITDA   90  BKR
39                      CC US Equity_PE_RATIO   75  CC 
40                      WY US Equity_PE_RATIO    9  WY 
41                      GM US Equity_PE_RATIO    9  GM 
42                      BK US Equity_PE_RATIO    6  BK 
43                      CI US Equity_PE_RATIO    6  CI 
44                     CMA US Equity_PE_RATIO    6  CMA
45                     KIM US Equity_PE_RATIO    3  KIM
46                     BKR US Equity_PE_RATIO   96  BKR
47      FO US Equity_DIVIDEND_INDICATED_YIELD   21  FO 
48      CC US Equity_DIVIDEND_INDICATED_YIELD   63  CC 
49      GM US Equity_DIVIDEND_INDICATED_YIELD   48  GM 
50     UFS US Equity_DIVIDEND_INDICATED_YIELD    3  UFS
51      CI US Equity_DIVIDEND_INDICATED_YIELD   12  CI 
52     AMD US Equity_DIVIDEND_INDICATED_YIELD  117  AMD
53               FO US Equity_EQY_DVD_YLD_IND   21  FO 
54               CC US Equity_EQY_DVD_YLD_IND   63  CC 
55               GM US Equity_EQY_DVD_YLD_IND   48  GM 
56              UFS US Equity_EQY_DVD_YLD_IND    3  UFS
57               CI US Equity_EQY_DVD_YLD_IND   12  CI 
58              AMD US Equity_EQY_DVD_YLD_IND  117  AMD
59                   HES US Equity_EPS_GROWTH    3  HES
60                  INTC US Equity_EPS_GROWTH    3  INT
61                   MOS US Equity_EPS_GROWTH    9  MOS
62                    FO US Equity_EPS_GROWTH    3  FO 
63                    CC US Equity_EPS_GROWTH   81  CC 
64                    GM US Equity_EPS_GROWTH    6  GM 
65                   HAS US Equity_EPS_GROWTH    3  HAS
66                   UFS US Equity_EPS_GROWTH    3  UFS
67                   AMD US Equity_EPS_GROWTH    3  AMD
68                    IR US Equity_EPS_GROWTH    3  IR 
69                   BKR US Equity_EPS_GROWTH  105  BKR

On remarque emaintenant qu'après avoir enlevé les deux denirères périodes,  nous n'avons plus que des valeurs manquantes de longues durées ou complètes (sur toutes les périodes soit 119). 
Nous allons procéder en 2 façons. 
Nous rappellons le but est de remplacer ou non les valeurs manquantes en n'atténuant pas l'effet des modalité présentes pour les indicateurs. On se dit par exmple qu'ici, remplacder par la moyenne ou médiane pourrait être intéressant, cependant nous craignions que cela affecte les tendances réelles des indicateurs. 

In [17]:
Valeurs_manquantes[Valeurs_manquantes["NA"]!=117]

index   NA  Nom
0            INTC US Equity_EBITDA_TO_REVENUE    3  INT
1             MOS US Equity_EBITDA_TO_REVENUE    9  MOS
2              CC US Equity_EBITDA_TO_REVENUE   48  CC 
4             HAS US Equity_EBITDA_TO_REVENUE    3  HAS
6             AMD US Equity_EBITDA_TO_REVENUE    3  AMD
8             BKR US Equity_EBITDA_TO_REVENUE   72  BKR
9      CC US Equity_CURRENT_EV_TO_T12M_EBITDA   63  CC 
11     GM US Equity_CURRENT_EV_TO_T12M_EBITDA    9  GM 
12    CMI US Equity_CURRENT_EV_TO_T12M_EBITDA    6  CMI
15    BKR US Equity_CURRENT_EV_TO_T12M_EBITDA   90  BKR
16  INTC US Equity_EBITDA_to_Interest_Expense   12  INT
17   MOS US Equity_EBITDA_to_Interest_Expense    9  MOS
18    CC US Equity_EBITDA_to_Interest_Expense   63  CC 
20   HAS US Equity_EBITDA_to_Interest_Expense    3  HAS
24   AMD US Equity_EBITDA_to_Interest_Expense    3  AMD
26   BKR US Equity_EBITDA_to_Interest_Expense   72  BKR
27   OXY US Equity_EBITDA_to_Interest_Expense   21  OXY
28          INTC US Equity_TOT_DEBT_TO_EBITDA    3  INT
29           MOS US Equity_TOT_DEBT_TO_EBITDA    9  MOS
30            CC US Equity_TOT_DEBT_TO_EBITDA   57  CC 
32            GM US Equity_TOT_DEBT_TO_EBITDA    3  GM 
33           CMI US Equity_TOT_DEBT_TO_EBITDA    6  CMI
34           HAS US Equity_TOT_DEBT_TO_EBITDA    3  HAS
36           AMD US Equity_TOT_DEBT_TO_EBITDA    3  AMD
38           BKR US Equity_TOT_DEBT_TO_EBITDA   90  BKR
39                      CC US Equity_PE_RATIO   75  CC 
40                      WY US Equity_PE_RATIO    9  WY 
41                      GM US Equity_PE_RATIO    9  GM 
42                      BK US Equity_PE_RATIO    6  BK 
43                      CI US Equity_PE_RATIO    6  CI 
44                     CMA US Equity_PE_RATIO    6  CMA
45                     KIM US Equity_PE_RATIO    3  KIM
46                     BKR US Equity_PE_RATIO   96  BKR
47      FO US Equity_DIVIDEND_INDICATED_YIELD   21  FO 
48      CC US Equity_DIVIDEND_INDICATED_YIELD   63  CC 
49      GM US Equity_DIVIDEND_INDICATED_YIELD   48  GM 
50     UFS US Equity_DIVIDEND_INDICATED_YIELD    3  UFS
51      CI US Equity_DIVIDEND_INDICATED_YIELD   12  CI 
53               FO US Equity_EQY_DVD_YLD_IND   21  FO 
54               CC US Equity_EQY_DVD_YLD_IND   63  CC 
55               GM US Equity_EQY_DVD_YLD_IND   48  GM 
56              UFS US Equity_EQY_DVD_YLD_IND    3  UFS
57               CI US Equity_EQY_DVD_YLD_IND   12  CI 
59                   HES US Equity_EPS_GROWTH    3  HES
60                  INTC US Equity_EPS_GROWTH    3  INT
61                   MOS US Equity_EPS_GROWTH    9  MOS
62                    FO US Equity_EPS_GROWTH    3  FO 
63                    CC US Equity_EPS_GROWTH   81  CC 
64                    GM US Equity_EPS_GROWTH    6  GM 
65                   HAS US Equity_EPS_GROWTH    3  HAS
66                   UFS US Equity_EPS_GROWTH    3  UFS
67                   AMD US Equity_EPS_GROWTH    3  AMD
68                    IR US Equity_EPS_GROWTH    3  IR 
69                   BKR US Equity_EPS_GROWTH  105  BKR

In [18]:
Valeurs_manquantes[Valeurs_manquantes["NA"]!=117]["NA"].sum()

1389

In [19]:
Valeurs_manquantes[Valeurs_manquantes["NA"]!=117]["Nom"].value_counts()

CC     8
BKR    6
GM     6
INT    4
HAS    4
AMD    4
MOS    4
CI     3
UFS    3
FO     3
CMI    2
BK     1
KIM    1
IR     1
HES    1
OXY    1
CMA    1
WY     1
Name: Nom, dtype: int64

Donc pour conserver l'effet des valeurs d'indicateurs déjà présentes, nous décidons de ne pas remplacer les valeurs manquantes. Nous proposons à la place d'établir une variable appelée "date_active" qui nous permettra de déterminer la date pour laquelle le titre aura tous ces indicateurs disponibles. C'est à partir de cette date ensuite que nous pourrons commencer à l'utiliser afin de prédire les rendements. Nous mettrons aussi une condition supplémentaire en plus: chacun de ces titres ne peut rentrer dans la modélisation qu' un an après sa date active (cela nous permet de posséder au moins 4 périodes d'indicateurs).

In [20]:
Valeurs_manquantes[Valeurs_manquantes["NA"]==117]

index   NA  Nom
3            WFC US Equity_EBITDA_TO_REVENUE  117  WFC
5             BK US Equity_EBITDA_TO_REVENUE  117  BK 
7            CMA US Equity_EBITDA_TO_REVENUE  117  CMA
10   WFC US Equity_CURRENT_EV_TO_T12M_EBITDA  117  WFC
13    BK US Equity_CURRENT_EV_TO_T12M_EBITDA  117  BK 
14   CMA US Equity_CURRENT_EV_TO_T12M_EBITDA  117  CMA
19  WFC US Equity_EBITDA_to_Interest_Expense  117  WFC
21   BK US Equity_EBITDA_to_Interest_Expense  117  BK 
22   MO US Equity_EBITDA_to_Interest_Expense  117  MO 
23  JLL US Equity_EBITDA_to_Interest_Expense  117  JLL
25  CMA US Equity_EBITDA_to_Interest_Expense  117  CMA
31          WFC US Equity_TOT_DEBT_TO_EBITDA  117  WFC
35           BK US Equity_TOT_DEBT_TO_EBITDA  117  BK 
37          CMA US Equity_TOT_DEBT_TO_EBITDA  117  CMA
52    AMD US Equity_DIVIDEND_INDICATED_YIELD  117  AMD
58             AMD US Equity_EQY_DVD_YLD_IND  117  AMD

In [21]:
Valeurs_manquantes[Valeurs_manquantes["NA"]==117]["Nom"].value_counts()

BK     4
WFC    4
CMA    4
AMD    2
JLL    1
MO     1
Name: Nom, dtype: int64

In [22]:
traitement = list(Valeurs_manquantes[Valeurs_manquantes["NA"]==117]["index"])

Ci-dessus, nous procédons d'une façon différente que précédemnt. Etant donné que la date active ne servirait à rien ici car elle nous amènerait à minima à la fin de l'étude et serait donc inutile et nous priverait de chacun de ces titres ce que nous ne souhaitons pas. Nous allons donc pour chacune de ces variables, remplacer les valeurs manquantes par -999. Cela nous permettra donc tous de même d'utiliser ces titres même si leurs données ne seront parfaites. Nous observons que les 3 titres CMA, WFC  et BK ont 4 indicateurs non disponibles soit près de 50% ce qui semble beaucoup. Nous les remplacons pour le moment mais nous réservons le droit de les enlever par la suite.

In [23]:
indicateurs.loc[0:116,:][traitement] = indicateurs.loc[0:116,:][traitement].fillna(value=-999)

C:\Users\louis\Anaconda3\lib\site-packages\pandas\core\frame.py:3494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [24]:
indicateurs#.info()

Dates  HES US Equity_EBITDA_TO_REVENUE  \
0   2010-01-31                         5.871767   
1   2010-02-28                         5.871767   
2   2010-03-31                         5.871767   
3   2010-04-30                         5.690633   
4   2010-05-31                         5.690633   
5   2010-06-30                         5.690633   
6   2010-07-31                         3.730067   
7   2010-08-31                         3.730067   
8   2010-09-30                         3.730067   
9   2010-10-31                         5.207067   
10  2010-11-30                         5.207067   
11  2010-12-31                         5.207067   
12  2011-01-31                         6.010767   
13  2011-02-28                         6.010767   
14  2011-03-31                         6.010767   
15  2011-04-30                         5.727533   
16  2011-05-31                         5.727533   
17  2011-06-30                         5.727533   
18  2011-07-31                         3.200600   
19  2011-08-31                         3.200600   
20  2011-09-30                         3.200600   
21  2011-10-31                         5.061833   
22  2011-11-30                         5.061833   
23  2011-12-31                         5.061833   
24  2012-01-31                        18.462233   
25  2012-02-29                        18.462233   
26  2012-03-31                        18.462233   
27  2012-04-30                        19.023267   
28  2012-05-31                        19.023267   
29  2012-06-30                        19.023267   
30  2012-07-31                        19.487833   
31  2012-08-31                        19.487833   
32  2012-09-30                        19.487833   
33  2012-10-31                         7.438500   
34  2012-11-30                         7.438500   
35  2012-12-31                         7.438500   
36  2013-01-31                        13.227433   
37  2013-02-28                        13.227433   
38  2013-03-31                        13.227433   
39  2013-04-30                        29.922900   
40  2013-05-31                        29.922900   
41  2013-06-30                        29.922900   
42  2013-07-31                        17.745100   
43  2013-08-31                        17.745100   
44  2013-09-30                        17.745100   
45  2013-10-31                        13.262300   
46  2013-11-30                        13.262300   
47  2013-12-31                        13.262300   
48  2014-01-31                        18.593333   
49  2014-02-28                        18.593333   
50  2014-03-31                        18.593333   
51  2014-04-30                        23.105933   
52  2014-05-31                        23.105933   
53  2014-06-30                        23.105933   
54  2014-07-31                        18.384367   
55  2014-08-31                        18.384367   
56  2014-09-30                        18.384367   
57  2014-10-31                        15.147967   
58  2014-11-30                        15.147967   
59  2014-12-31                        15.147967   
60  2015-01-31                         6.545300   
61  2015-02-28                         6.545300   
62  2015-03-31                         6.545300   
63  2015-04-30                         7.219667   
64  2015-05-31                         7.219667   
65  2015-06-30                         7.219667   
66  2015-07-31                        10.273300   
67  2015-08-31                        10.273300   
68  2015-09-30                        10.273300   
69  2015-10-31                       -25.486200   
70  2015-11-30                       -25.486200   
71  2015-12-31                       -25.486200   
72  2016-01-31                         3.391567   
73  2016-02-29                         3.391567   
74  2016-03-31                         3.391567   
75  2016-04-30                         4.330067   
76  2016-05-31                         4.330067   
77  2016-06-30                 

In [25]:
pos=0
Ticker_abr=[]
Dates = []
EBITDA_TO_REVENUE = []
CURRENT_EV_TO_T12M_EBITDA = []
EBITDA_to_Interest_Expense=[]
TOT_DEBT_TO_EBITDA =[]
BEST_EST_PE_NXT_YR2 = []
PE_RATIO = []
DIVIDEND_INDICATED_YIELD = []
EQY_DVD_YLD_IND = []
EPS_GROWTH = []
while pos < 50 : 
        a=[x for x in list(indicateurs.columns) if Equivalence['Abréviation'][pos] in x.upper()]
        a.append("Dates")
        table = indicateurs[a]
        #x=Equivalence['Abréviation'][pos]
        x = [Equivalence['Abréviation'][pos]] * len(table)
        Ticker_abr.extend(x)
        Dates.extend(pd.to_datetime(table.iloc[:,9]))
        #Dates.extend(table.iloc[:,8].astype(str))
        EBITDA_TO_REVENUE.extend(table.iloc[:,0])#.shift(1) #comme annonce de marché 
        CURRENT_EV_TO_T12M_EBITDA.extend(table.iloc[:,1])#.shift(1)
        EBITDA_to_Interest_Expense.extend(table.iloc[:,2])#.shift(1)
        TOT_DEBT_TO_EBITDA.extend(table.iloc[:,3])#.shift(1)
        BEST_EST_PE_NXT_YR2.extend(table.iloc[:,4])
        PE_RATIO.extend(table.iloc[:,5])#.shift(1)
        DIVIDEND_INDICATED_YIELD.extend(table.iloc[:,6])#.shift(1)
        EQY_DVD_YLD_IND.extend(table.iloc[:,7])#.shift(1)
        EPS_GROWTH.extend(table.iloc[:,8])#.shift(1)
        pos+=1
Indicateur_titres = pd.DataFrame(list(zip(Ticker_abr,Dates,EBITDA_TO_REVENUE,CURRENT_EV_TO_T12M_EBITDA,EBITDA_to_Interest_Expense,TOT_DEBT_TO_EBITDA,BEST_EST_PE_NXT_YR2,PE_RATIO,DIVIDEND_INDICATED_YIELD,EQY_DVD_YLD_IND,EPS_GROWTH)),columns=["Ticker_abr","Dates","EBITDA_TO_REVENUE","CURRENT_EV_TO_T12M_EBITDA","EBITDA_to_Interest_Expense","TOT_DEBT_TO_EBITDA","BEST_EST_PE_NXT_YR2","PE_RATIO","DIVIDEND_INDICATED_YIELD","EQY_DVD_YLD_IND","EPS_GROWTH"])

Nous avons crée ici une table suivant le même format que la table des rendements créée précédement. Nous avons donc inscrit chaque indicateur en fonction du titre et de son mois. 
Cette façon de procéder nous permettra de venir plugger ces indicateurs à notre table des rendements. Aussi, nous avons mis tous ces indicateurs au mois et non plus au trimestre par souci pratique.

In [26]:
Indicateur_titres.to_csv("db_indicateurs.csv", header=True,index=False)

In [27]:
Indicateur_titres

Ticker_abr      Dates  EBITDA_TO_REVENUE  CURRENT_EV_TO_T12M_EBITDA  \
0        HES US 2010-01-31           5.871767                   1.529033   
1        HES US 2010-02-28           5.871767                   1.529033   
2        HES US 2010-03-31           5.871767                   1.529033   
3        HES US 2010-04-30           5.690633                   1.170033   
4        HES US 2010-05-31           5.690633                   1.170033   
...         ...        ...                ...                        ...   
5945     CCL US 2019-07-31           7.236967                   2.396533   
5946     CCL US 2019-08-31           7.236967                   2.396533   
5947     CCL US 2019-09-30           7.236967                   2.396533   
5948     CCL US 2019-10-31                NaN                        NaN   
5949     CCL US 2019-11-30                NaN                        NaN   

      EBITDA_to_Interest_Expense  TOT_DEBT_TO_EBITDA  BEST_EST_PE_NXT_YR2  \
0                       6.472233              0.2808             -999.000   
1                       6.472233              0.2808             -999.000   
2                       6.472233              0.2808             -999.000   
3                       5.301200              0.2575             -999.000   
4                       5.301200              0.2575             -999.000   
...                          ...                 ...                  ...   
5945                    7.061233              0.7053                2.962   
5946                    7.061233              0.7053                2.962   
5947                    7.061233              0.7053                2.962   
5948                         NaN                 NaN                  NaN   
5949                         NaN                 NaN                  NaN   

      PE_RATIO  DIVIDEND_INDICATED_YIELD  EQY_DVD_YLD_IND  EPS_GROWTH  
0     6.019033                  0.213167         0.213167         NaN  
1     6.019033                  0.213167         0.213167         NaN  
2     6.019033                  0.213167         0.213167         NaN  
3     3.655333                  0.264867         0.264867   90.322567  
4     3.655333                  0.264867         0.264867   90.322567  
...        ...                       ...              ...         ...  
5945  3.247033                  1.525200         1.525200   -4.225367  
5946  3.247033                  1.525200         1.525200   -4.225367  
5947  3.247033                  1.525200         1.525200   -4.225367  
5948       NaN                       NaN              NaN         NaN  
5949       NaN                       NaN              NaN         NaN  

[5950 rows x 11 columns]

In [28]:
Indicateur_titres.isna().sum()

Ticker_abr                      0
Dates                           0
EBITDA_TO_REVENUE             238
CURRENT_EV_TO_T12M_EBITDA     268
EBITDA_to_Interest_Expense    283
TOT_DEBT_TO_EBITDA            274
BEST_EST_PE_NXT_YR2           100
PE_RATIO                      310
DIVIDEND_INDICATED_YIELD      247
EQY_DVD_YLD_IND               247
EPS_GROWTH                    322
dtype: int64

In [29]:
238+268+283+274+100+310+247+247+322  #900 de plus car on ne comptabilisais pas la NA liés au deux derniers mois 
# soient 50*2*9=900

2289

# Table Activité

In [34]:
def départ(df):
    titre = []
    début = []
    for company in list(df.columns[1:]):
        ent = company
        start = df[["Fonds","{}".format(ent)]].dropna()["Fonds"].min()
        if start > pd.to_datetime('2010-01-31 00:00:00') : 
            titre.append(company)
            début.append(start)
        départ_différé = pd.DataFrame(list(zip(titre,début)),columns=["titre","start"])
    return départ_différé
fusion =départ(IND1)
fusion

titre      start
0   CC US Equity 2013-04-30
1  BKR US Equity 2016-01-31

In [35]:
fusion =départ(IND1)
for tb in [IND2,IND3,IND4,IND6,IND7,IND8,IND9]:  #IND5
    tb2 = départ(tb)
    fusion = fusion.merge(tb2,how = "outer",left_on="titre",right_on="titre")
fusion.columns=["Tickers","EBITDA_TO_REVENUE","CURRENT_EV_TO_T12M_EBITDA","EBITDA_to_Interest_Expense","TOT_DEBT_TO_EBITDA","PE_RATIO","DIVIDEND_INDICATED_YIELD","EQY_DVD_YLD_IND","EPS_GROWTH"]
#,"BEST_EST_PE_NXT_YR"

#fusion =fusion.fillna(pd.to_datetime("1999-01-01"))
fusion["Date_active"] = fusion.loc[:,"EBITDA_TO_REVENUE":"EPS_GROWTH"].values.max(axis=1)
#fusion[["Tickers","Date_active"]]# La date à laquelle chacun de ces titres rentrera dans notre portefeuille.
fusion.head(4)

Tickers EBITDA_TO_REVENUE CURRENT_EV_TO_T12M_EBITDA  \
0   CC US Equity        2013-04-30                2015-04-30   
1  BKR US Equity        2016-01-31                2017-07-31   
2   GM US Equity               NaT                2010-10-31   
3  CMI US Equity               NaT                2010-07-31   

  EBITDA_to_Interest_Expense TOT_DEBT_TO_EBITDA   PE_RATIO  \
0                 2014-07-31         2014-01-31 2016-04-30   
1                 2016-01-31         2017-07-31 2018-01-31   
2                        NaT         2010-04-30 2010-10-31   
3                        NaT         2010-07-31        NaT   

  DIVIDEND_INDICATED_YIELD EQY_DVD_YLD_IND EPS_GROWTH Date_active  
0               2015-04-30      2015-04-30 2016-01-31  2016-04-30  
1                      NaT             NaT 2018-10-31  2018-10-31  
2               2014-01-31      2014-01-31 2010-07-31  2014-01-31  
3                      NaT             NaT        NaT  2010-07-31

In [36]:
fusion[["Tickers","Date_active"]]

Tickers Date_active
0     CC US Equity  2016-04-30
1    BKR US Equity  2018-10-31
2     GM US Equity  2014-01-31
3    CMI US Equity  2010-07-31
4   INTC US Equity  2010-10-31
5    OXY US Equity  2011-10-31
6     WY US Equity  2010-10-31
7     BK US Equity  2010-07-31
8     CI US Equity  2011-01-31
9    CMA US Equity  2010-07-31
10   KIM US Equity  2010-04-30
11    FO US Equity  2011-10-31
12   UFS US Equity  2010-04-30
13   HES US Equity  2010-04-30
14    IR US Equity  2010-04-30

In [37]:
fusion.to_csv("Start_indicateurs.csv",header=True,index=False)

# Table d'entraînement 

In [38]:
print(Indicateur_titres.shape) # A vérifier, on devrait normalement avoir le meme nombre d'indicateur que de rendements
# car c'est indexé sur les dates.
print(Equivalence.shape)
print(db_rendements.shape)

(5950, 11)
(50, 3)
(5875, 7)


In [39]:
Train_Dataset = Equivalence.merge(Indicateur_titres,how="left", left_on="Abréviation",right_on="Ticker_abr")
del Train_Dataset["Ticker_abr"]
Train_Dataset = db_rendements.merge(Train_Dataset,how="left", left_on=["Date","Titres"],right_on=["Dates","Name"])
del Train_Dataset["Name"]
del Train_Dataset["Dates"]
del Train_Dataset["Abréviation"]
temp = fusion[["Tickers","Date_active"]]
Train_Dataset = Train_Dataset.merge(temp,how="left",left_on="Ticker",right_on="Tickers")
del Train_Dataset["Tickers"]
Train_Dataset["Date_active"] = Train_Dataset["Date_active"].fillna(pd.to_datetime("1997-04-24"))
Train_Dataset.shape#On rajoute une année à la date  d'entrée en activité du titre pour avoir plusieurs trimestre d'indicateur.

(5875, 18)

In [40]:
# Suppression des titres aux périodes où nous ne les utiliserons pas :
Train_setF = Train_Dataset[Train_Dataset["Date"] > Train_Dataset["Date_active"]]
#Train_setF["Date_active2"] = Train_setF["Date_active"].apply(lambda x: x.replace(year=x.year + 1))#.fillna(pd.to_datetime("1997-04-24"))
#Train_setF=Train_setF[Train_setF["Date"]>=Train_setF["Date_active2"]]
#del Train_setF["Date_active"]
#del Train_setF["Date_active2"]
#Train_setF['year'] = Train_setF['Date'].dt.year
Train_setF['month'] = Train_setF['Date'].dt.month
Train_setF["période"] = np.where(Train_setF["month"].isin([1,4,7,10]),1,np.where(Train_setF["month"].isin([2,5,8,11]),2,np.where(Train_setF["month"].isin([3,6,9,12]),3,None)))
Train_setF.shape

C:\Users\louis\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\louis\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


(5605, 20)

In [41]:
Train_setF

Date                     Titres     Return      lag1      lag2  \
1    2010-01-31      NEWMONT GOLDCORP CORP  -9.406045       NaN       NaN   
3    2010-01-31  INTL FLAVORS & FRAGRANCES  -3.330092       NaN       NaN   
4    2010-01-31              MOSAIC CO/THE -10.413530       NaN       NaN   
5    2010-01-31                EXELON CORP  -6.650297       NaN       NaN   
6    2010-01-31        KIMBERLY-CLARK CORP  -6.780725       NaN       NaN   
...         ...                        ...        ...       ...       ...   
5870 2019-11-30  OCCIDENTAL PETROLEUM CORP  -4.765432 -8.927367  4.073635   
5871 2019-11-30           WILLIAMS COS INC   1.837741 -7.273483  3.544075   
5872 2019-11-30         UNION PACIFIC CORP   6.946350  2.148413  0.012349   
5873 2019-11-30       WASTE MANAGEMENT INC   0.623830 -2.426087 -3.230081   
5874 2019-11-30              CARNIVAL CORP   6.327681 -1.876001 -0.839383   

           lag3       lag4         Ticker  EBITDA_TO_REVENUE  \
1           NaN        NaN  NEM US Equity          16.904533   
3           NaN        NaN  IFF US Equity           6.352100   
4           NaN        NaN  MOS US Equity          11.826033   
5           NaN        NaN  EXC US Equity          15.841000   
6           NaN        NaN  KMB US Equity           5.915200   
...         ...        ...            ...                ...   
5870 -15.342680   2.147971  OXY US Equity                NaN   
5871  -4.220779 -12.125540  WMB US Equity                NaN   
5872  -9.457417   6.410029  UNP US Equity                NaN   
5873   2.008547   1.412846   WM US Equity                NaN   
5874  -5.634554   1.460795  CCL US Equity                NaN   

      CURRENT_EV_TO_T12M_EBITDA  EBITDA_to_Interest_Expense  \
1                      2.190000                    4.858967   
3                      3.541267                    3.261400   
4                      6.835100                   13.253000   
5                      1.718900                    3.861567   
6                      2.934567                    4.688533   
...                         ...                         ...   
5870                        NaN                         NaN   
5871                        NaN                         NaN   
5872                        NaN                         NaN   
5873                        NaN                         NaN   
5874                        NaN                         NaN   

      TOT_DEBT_TO_EBITDA  BEST_EST_PE_NXT_YR2   PE_RATIO  \
1               0.354567               -999.0   4.947800   
3               0.763733               -999.0   5.414900   
4               0.261067               -999.0  30.601933   
5               0.566600               -999.0   3.738633   
6               0.490567               -999.0   4.130667   
...                  ...                  ...        ...   
5870                 NaN                  NaN        NaN   
5871                 NaN                  NaN        NaN   
5872                 NaN                  NaN        NaN   
5873                 NaN                  NaN        NaN   
5874                 NaN                  NaN        NaN   

      DIVIDEND_INDICATED_YIELD  EQY_DVD_YLD_IND  EPS_GROWTH Date_active  \
1                     0.261800         0.261800   59.166667  1997-04-24   
3                     0.699267         0.699267   11.111100  1997-04-24   
4                     0.109700         0.109700   56.565667  1997-04-24   
5                     1.597800         1.597800    1.543200  1997-04-24   
6                     1.460200         1.460200   -2.040800  1997-04-24   
...                        ...              ...         ...         ...   
5870                       NaN              NaN         NaN  2011-10-31   
5871                       NaN              NaN         NaN  1997-04-24   
5872                       NaN              NaN         NaN  1997-04-24   
5873                       NaN              NaN         NaN  1997-04-24   
5874          

In [30]:
Train_setF.to_csv("db_train.csv",header=True,index=False)

##### FIN 